tidy3d/components/validators.py

```diff
@@ -396,3 +396,21 @@ def freqs_not_empty(cls, val):
         return val
 
     return freqs_not_empty
+
+
+def validate_mode_plane_radius(mode_spec: "ModeSpec", plane: "Box", msg_prefix: str):  # noqa: F821
```

Comment:
I think it's fine (and perhaps better) to import these types and use directly here.

Reply:
That would be circular import as usually validators are imported in other components rather than the other way round. However, I don't know if this is the best place to put this function: it seems that everything else in this file creates an actual pydantic validator, while this is just a convenient method to use in a post-init validator.

Reply:
i think validators already imports from `Box` https://github.com/flexcompute/tidy3d/blob/develop/tidy3d/components/validators.py#L10 and ModeSpec doesnt import from validators, so is it circular?

Reply:
Here's my thought:

* if the type annotation causes circular imports, we probably should be rethinking where to place the function in the first place. So in this case, should this just go in simulation.py for example? what's the benefit of separating it into validators.py when it just gets used in simulation things?
* If we want to keep it validators.py just for some separation, I think it's probably ok to import Box and ModeSpec at least the way it looks now, but it does mean we can't import validators from here those modules in the future.

so maybe overall I think it makes a bit more sense to just put this in `simulation.py`? or nearby (eg in new module `base_sim/validators.py`?)

Reply:
> i think validators already imports from `Box` https://github.com/flexcompute/tidy3d/blob/develop/tidy3d/components/validators.py#L10 and ModeSpec doesnt import from validators, so is it circular?

Oh interesting, sorry I was just thinking about it conceptually that this validators file would make sense to me to be at the top of the hierarchy. I'm not sure if there's any logic to it right now, it seems to me like we *could* at some point have a validator we may want to import in Box or ModeSpec and it just so happens that we don't have one now?

> in any case, while separating the validators into their own file was maybe not a bad idea, functionally it's probably fine if this validator was just defined inside of simulation.py or wherever its used.

So I put it here because I need to use it both in Simulation and in ModeSolver. Any better ideas?

Reply:
> Oh interesting, sorry I was just thinking about it conceptually that this validators file would make sense to me to be at the top of the hierarchy.

Yea i'm also thinking about this out loud because I think it was never super well defined. Conceptually that makes sense, but also if these validators are going to use specific information about Box or ModeSpec components , they probably shouldnt be here. Maybe these more specific validators need to just go where they are used.

> So I put it here because I need to use it both in Simulation and in ModeSolver.

Yea maybe in `simulation.py` or actually a new file `base_sim/validators.py` makes some sense? the ModeSolver would import from `base_sim` right?

Reply:
I think the original idea of the `validators.py` file is exactly for things that need to be imported in multiple files. Generally it makes sense to me that there may be such validators but not sure if a more fine-grained splitting would make more sense.

Reply:
yea it's just that there's some conflict between: being able to import these validators in many files vs. these validators being able to know about things in said files (eg. ModeSpec).

For now maybe my preference is make a new validators.py in `sim_base` and put this there (+ maybe other simulation specific validators that live in the current file?

Reply:
So I started doing this but actually I don't understand how this helps clarify the structure. What is the logic/difference between having validators in `components/validators.py` and `components/base_sim/validators.py`?

Reply:
Updated.